### ISSUE TITLE
Using super_annotations with source_gen

### ISSUE THREAD

**Issue by bluegearapps** [schultek/super_annotations#2]
I am using source_gen to generate other .g.dart files (for model generation and json serialization). Now I am adding super_annotations for static metaprogramming. When I run the command it throws error:

```
Both source_gen:combining_builder and super_annotations:super_annotations may output lib/main_libra_cart.g.dart. Potential outputs must be unique across all builders. See https://github.com/dart-lang/build/blob/master/docs/faq.md#why-do-builders-need-unique-outputs
```
Based on my understanding, following the guide above, I have to add new build.yaml config, with `generate_for`

From source_gen:
```
The combining builder reads in all the pieces written by different shared part builders and writes them to the final .g.dart output in the user's source directory. You should never use the .g.dart extension for any other Builder.
```

How to modify the ouput so the don't have conflict? Thank you!

**Comment by schultek** [comment]
I published a fix in version `0.3.7` for this.
To resolve the conflict, you have to define a custom target for `super_annotations` in your `build.yaml`:
```yaml
global_options:
  super_annotations:
    options:
      targets: [ mytarget ]
```

This will then generate `main.mytarget.dart` instead of `main.g.dart` and therefore should not have any conflict with source_gen.

**Comment by bluegearapps** [comment]
Thank you so much for your answer, I have a build.yaml only includes differrent files for different builders, and that resolved the issue

```
targets:
  $default:
    builders:
      built_value_generator|built_value:
        generate_for:
          include:
            - lib/models/**.dart
      source_gen|combining_builder:
        generate_for:
          include:
            - lib/models/**.dart
      super_annotations:
        generate_for:
          include:
            - lib/display_models/**.dart
```
    
I have one class `DisplayAnnotation` extends `ClassAnnotation`
And use the annotation in 3 other classes, using mixin, very similar to this example https://github.com/schultek/super_annotations/tree/main/examples/part_of_strategy

`class_a.dart` file:
```dart
@CodeGen(runAfter: [CodeGen.addPartOfDirective])
library test1;

part 'class_a.g.dart';

class ClassA extends StatefulWidget{
  @override
  State<StatefulWidget> createState() {
    return ClassAState();
  }
}

@ DisplayAnnotation(
    name: 'Test1',
)
class ClassAState extends State<ClassA> with _$ClassAState{
  @override
  void initState(){
    generatedInitState();
    super.initState();
}

```
and `class_b.dart` file:

```dart
@CodeGen(runAfter: [CodeGen.addPartOfDirective])
library test2;

part 'class_b.g.dart';

class ClassB extends StatefulWidget{
  @override
  State<StatefulWidget> createState() {
    return ClassBState();
  }
}

@ DisplayAnnotation(
    name: 'Test2',
)
class ClassBState extends State<ClassB> with _$ClassBState{
  @override
  void initState(){
    generatedInitState();
    super.initState();
}

```
and `class_c.dart` file:

```dart
@CodeGen(runAfter: [CodeGen.addPartOfDirective])
library test3;

part 'class_c.g.dart';

class ClassC extends StatefulWidget{
  @override
  State<StatefulWidget> createState() {
    return ClassCState();
  }
}

@ DisplayAnnotation(
    name: 'Test3',
)
class ClassCState extends State<ClassC> with _$ClassCState{
  @override
  void initState(){
    generatedInitState();
    super.initState();
}

```

After running the build command: `flutter packages pub run build_runner build --delete-conflicting-outputs`
The .g.dart files are generated, each file *has all the contents* of other files: class_a.g.dart has content of class_b.g.dart and class_c.g.dart, also class_b.g.dart has all content of class_a.g.dart and class_c.g.dart,

Do you have any idea about this issue?

**Comment by schultek** [comment]
Does your files import each other in any way?

**Comment by bluegearapps** [comment]
Right!!! I found the in-direct import in the other files, that is the reason. I can resolve this by adding the import manually via 
```dart
  output.directives.add(Directive.import('package:my_package/my_file.dart'));
```
Thank you so much for the great package, very smart way to generate source code!

**Comment by bluegearapps** [comment]
Sorry, after adding the import to annotation's apply method:

```dart
@override
  void apply(Class target, LibraryBuilder output) {
    output.directives.addAll(
        [Directive.import('package: my_package/my_file1.dart'),
         Directive.import('package: my_package/my_file2.dart')
        ]
    );

    output.body.add(Mixin((e) => e
      ..name = '_\$${target.name}'
      ..fields.addAll([
....

```

then I got the error:
```bash
[SEVERE] super_annotations:super_annotations on lib/display_models/class_a.dart:

Could not format because the source could not be parsed:

line 1, column 353 of .: The part-of directive must be the only directive in a part.
  ╷
1 │ import 'package: my_package/my_file1.dart';import 'package: my_package/my_file2.dart';part of 'class_a.dart';mixin _$ClassAState  {
```

Do you have any idea on this? It looks correct because the series of imports are added on top of the file, then `part of`, then class code...

**Comment by bluegearapps** [comment]
I just found that I cannot put those imports into apply method. So where is the best place to put the import which has cross-reference to other generated lib? Thank you

**Comment by schultek** [comment]
You cannot have imports in a part-of file in dart. 
Are your classes all dependent on each other?
You could maybe put those in the same library (via part-of) or make a new file that imports them all and only activate code gen for this one file.

**Comment by schultek** [comment]
I'm thinking about adding a flag that controls whether it recursively searches all imports for annotated classes (how it is now) or just consideres the starting library. 

Alternatively a setting to ignore specific classes for code generation. 

That would help, right?

**Comment by bluegearapps** [comment]
Thank you, I think just consider the starting library would help.

**Comment by schultek** [comment]
I published version 0.4.0. 
You can now do `@CodeGen(discoveryMode: DiscoveryMode.inputLibrary)`

**Comment by bluegearapps** [comment]
Thanks @schultek , it's working now. 
Merry Christmas! 🎄
